File: homing-studio-base/src/main/java/hue/captains/singapura/js/homing/studio/base/ui/MasterDetail.java
package hue.captains.singapura.js.homing.studio.base.ui;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;
import hue.captains.singapura.js.homing.core.ModuleImports;
import hue.captains.singapura.js.homing.core.js.TreeRendererModule;

import java.util.List;

/**
 * A tree on the left, whatever the selected row is about on the right — built
 * once and used by both surfaces that want it.
 *
 * <p>The catalogue listing and the theme picker had arrived at the same shape
 * independently: create a nav element, create a detail element, construct a
 * {@code TreeRenderer} over the nav, and repaint the detail from
 * {@code onSelect}. Same fifteen lines, two copies, and they had already drifted
 * — one put the detail first and divided by the golden ratio, the other put the
 * tree first and sized it to content.</p>
 *
 * <p>What differs between them is only what the detail draws, which is the
 * caller's business and stays there: {@code onSelect} receives the selection and
 * the body element, and does as it likes with both.</p>
 *
 * <p>The tree is on the LEFT in both, now. Table of contents beside content is
 * the pattern a reader already knows, and the reason the catalogue did not do it
 * before — not knowing how wide a tree wanted to be — is answered by
 * {@link MasterDetailStyles#md_nav()} sizing to its content.</p>
 */
public record MasterDetail() implements DomModule<MasterDetail> {

    /** Builds the pair into a host and returns { navEl, bodyEl, renderer }. */
    public record mountMasterDetail() implements Exportable._Constant<MasterDetail> {}

    public static final MasterDetail INSTANCE = new MasterDetail();

    @Override
    public ImportsFor<MasterDetail> imports() {
        return ImportsFor.<MasterDetail>builder()
                .add(new ModuleImports<>(List.of(new TreeRendererModule.TreeRenderer()),
                        TreeRendererModule.INSTANCE))
                .add(new ModuleImports<>(List.of(
                        new MasterDetailStyles.md_split(),
                        new MasterDetailStyles.md_nav(),
                        new MasterDetailStyles.md_body()
                ), MasterDetailStyles.INSTANCE))
                .build();
    }

    @Override
    public ExportsOf<MasterDetail> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new mountMasterDetail()));
    }
}

File: homing-studio-base/src/main/resources/homing/js/hue/captains/singapura/js/homing/studio/base/ui/MasterDetail.js
// MasterDetail.js
//
// A tree on the LEFT, whatever the selected row is about on the RIGHT.
//
//   mountMasterDetail(opts) -> { navEl, bodyEl, renderer }
//
// The catalogue listing and the theme picker had arrived at the same fifteen
// lines independently — make a nav, make a detail, build a TreeRenderer over the
// nav, repaint the detail from onSelect — and had already drifted: one put the
// detail first and divided by the golden ratio, the other put the tree first and
// sized it to content. This is the one copy.
//
// What differs between them is only what the detail DRAWS, which stays with the
// caller: onSelect receives the selection and the body element and does as it
// likes with both.
//
// The nav is sized by its content (see MasterDetailStyles.md_nav), which is what
// makes table-of-contents-on-the-left workable. Not knowing how wide a tree
// wanted to be is why the catalogue used a ratio instead; that is answered now.
//
// The framework's EsModuleWriter appends the import/export prologue from the
// matching Java MasterDetail declaration — do not add import/export lines here.

var _mdSeq = 0;

/**
 * opts:
 *   branch       (required) DomOpsParty branch — the caller owns the lifetime
 *   host         (required) element to append the pair into
 *   data         (required) tree payload for TreeRenderer.setData
 *   onSelect     (bodyEl, sel) => void   — repaint the detail
 *   onActivate   (sel) => void           — Enter / double-click
 *   expandDepth, showBadge, showNote, showRoot, hrefForPath — passed through
 *
 * Returns { navEl, bodyEl, renderer }. The caller forwards keydown: the host
 * owns WHEN keys flow, the renderer owns what they mean.
 */
function mountMasterDetail(opts) {
    if (!opts || !opts.branch) throw new Error("mountMasterDetail: opts.branch is required");
    if (!opts.host)            throw new Error("mountMasterDetail: opts.host is required");
    if (!opts.data)            throw new Error("mountMasterDetail: opts.data is required");

    var branch = opts.branch;
    var seq = ++_mdSeq;

    var split = branch.createElement("mdsplit" + seq, "div");
    css.addClass(split, md_split);

    // Nav first: table of contents beside content is the order a reader knows,
    // and it is the reading order too.
    var navEl = branch.createElement("mdnav" + seq, "div");
    css.addClass(navEl, md_nav);
    navEl.setAttribute("tabindex", "0");
    split.appendChild(navEl);

    var bodyEl = branch.createElement("mdbody" + seq, "div");
    css.addClass(bodyEl, md_body);
    split.appendChild(bodyEl);

    opts.host.appendChild(split);

    var onSelect = opts.onSelect || function () {};
    var onActivate = opts.onActivate || function () {};

    var renderer = new TreeRenderer({
        branch:      branch,
        container:   navEl,
        expandDepth: (typeof opts.expandDepth === "number") ? opts.expandDepth : 0,
        showBadge:   !!opts.showBadge,
        showNote:    !!opts.showNote,
        showRoot:    !!opts.showRoot,
        hrefForPath: opts.hrefForPath || null,
        onSelect:    function (sel) { onSelect(bodyEl, sel); },
        onActivate:  function (sel) { onActivate(sel); }
    });
    renderer.setData(opts.data);

    return { splitEl: split, navEl: navEl, bodyEl: bodyEl, renderer: renderer };
}
